File: LCA.java

import java.util.ArrayList;
import java.util.List;

class Node {
	int data;
	Node left, right;

	Node(int value) {
		data = value;
		left = right = null;
	}
}

public class LCA {

	Node root;
	private List<Integer> path1 = new ArrayList<>();
	private List<Integer> path2 = new ArrayList<>();

	int findLCA(int a, int b) {
		path1.clear();
		path2.clear();
		return findLowestCommonAncestor(root, a, b);
	}

	private int findLowestCommonAncestor(Node root, int a, int b) {
		int i;
		if (!findPath(root, a, path1) || !findPath(root, b, path2)) {

			if(path1.size()==0)
			{
				System.out.println("First number is missing");
			}
			else if (path2.size()== 0)
			{
				System.out.println("Second number is missing");
			}
			return -1;
		}

		for (i = 0; i < path1.size() && i < path2.size(); i++) {
			if (!path1.get(i).equals(path2.get(i)))
				return path1.get(i-1);
		}
		return path1.get(i-1);
	}

	private boolean findPath(Node root, int val, List<Integer> path)
	{
		if (root == null) {
			return false;
		}

		path.add(root.data);

		if (root.data == val) {
			return true;
		}

		if (root.left != null && findPath(root.left, val, path)) {
			return true;
		}

		if (root.right != null && findPath(root.right, val, path)) {
			return true;
		}

		path.remove(path.size()-1);
		return false;
	}

	public static void main(String[] args) {

		LCA bt = new LCA();
		bt.root = new Node(1);
		bt.root.right = new Node(2);
		bt.root.left = new Node(3);
		bt.root.right.right = new Node(4);
		bt.root.right.left = new Node(5);
		bt.root.left.right = new Node(6);
		bt.root.left.left = new Node(7);

		System.out.println("LCA of 4 and 5 is " + bt.findLCA(4,5));
		System.out.println("LCA of 6 and 7 is " + bt.findLCA(6,7));
		System.out.println("LCA of 2 and 4 is " + bt.findLCA(2,4));
		System.out.println("LCA of 7 and 5 is " + bt.findLCA(7,5));
		System.out.println("LCA of 7 and 9 is " + bt.findLCA(7,9));
		

	}
}

File: LCA.py

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def findLCA(root, a, b):
    path1 = []
    path2 = []

    if (not findPath(root, a, path1) or not findPath(root, b,path2)):
        return -1

    i = 0
    while (i < len(path1) and i < len(path2)):
        if path1[i] != path2[i]:
            return path1[i - 1]
        i += 1
    return path1[i - 1]

def findPath(root, val, path):

    path.append(root.data)

    if root.data == val:
        return True

    if ((root.left != None and findPath(root.left, val, path)) or
            (root.right != None and findPath(root.right, val,path))):
        return True

    path.pop()
    return False

root = Node(1)
root.right = Node(2)
root.left = Node(3)
root.right.right = Node(4)
root.right.left = Node(5)
root.left.right = Node(6)
root.left.left = Node(7)

